doc/dev/static_type_checking.md

```diff
@@ -0,0 +1,1399 @@
+our libraries.
+
+### Use typing.Union when accepting more than one type
+
+[typing.Union](https://docs.python.org/3/library/typing.html#typing.Union) allows for typing something with more than
+one type. For example, a Union[str, DocumentBuildMode] means that either `str` or `DocumentBuildMode` are permissible for
+the `build_mode` parameter:
+
+```python
+from typing import Union
+
+
+def begin_build_model(
+    self, source: str, build_mode: Union[str, DocumentBuildMode], **kwargs: Any
+) -> DocumentModelAdministrationLROPoller[DocumentModelInfo]:
+    ...
+```
+
+A `Union` requires at least two types and should be used with type that are not consistent with each other. For
+example, usage of `Union[int, float]` is not necessary since `int` is consistent with `float` (implements all the supported operations of `float`) -- you can just use `float`. It's
+also recommended avoiding having functions return `Union` types as it causes the user to need to understand or perform `isinstance` checks
+on the return value before acting on it to write type-safe code. Sometimes this can be resolved by using an [overload](#use-typingoverload-to-overload-a-function).
+
+
+### Use typing.Optional when a parameter can explicitly be None
+
+Python's typing docs for [Optional](https://docs.python.org/3/library/typing.html#typing.Optional) explain that
+
+> Optional[X] is equivalent to X | None (or Union[X, None])
+
+Optional is just a specialized Union and somewhat confusingly named. `Optional` should only be used when an argument can be either
+something _or_ `None`. It should not be used just because a parameter has a default argument. For example, `Optional` usage is appropriate here:
+
+```python
+from typing import Optional, Any
+
+def begin_export_project(
+    project_name: str,
+    *,
+    string_index_type: str = "UnicodeCodePoint",  # Param with a default, doesn't allow None, doesn't require Optional
+    asset_kind: Optional[str] = None,  # None is allowed, type as Optional
+    **kwargs: Any
+) -> LROPoller[JSON]:
+    ...
+```
+
+Per PEP 484, type checkers have moved towards requiring the Optional type to be made explicit when a default of None is provided.
+
+> Note: The `X | None` syntax is only supported in Python 3.10+.
+
+### Typing for collections
+
+A few things to think about when typing collections in Python.
+
+1) "Be conservative in what you send, be liberal in what you accept." - The Robustness Principle
+
+   In other words, we should aim to be lenient in the types we accept as a parameter because this allows for more
+   flexibility for the caller, e.g. accepting an `Iterable` (where appropriate) allows for more possible types to be
+   passed (list, tuple, dict, etc.) than just specifying a `List`. On the other hand, we should aim to be specific in
+   what we say we return - it's more helpful to specify concrete types which a user can reason about than abstract types.
```

Comment:
This is true, however for APIs that we expect to grow over time, a more abstract return type gives us more flexibility.
For example typing the return value as `Iterable` rather than `List` means we can develop the returned object so long as it adheres to the protocol.

Reply:
Yeah I see what you mean. I think what's written is a good rule of thumb, but maybe not so great for long-lived APIs which cannot have breaking changes. Updating!